### ISSUE TITLE
Does IBCMerge make sense for crossgen?

### ISSUE THREAD

**Issue by ericstj** [dotnet/runtime#34422]
Today when we build libraries we embed some version of IBC data in the binary, only to later read that during the installer phase when cross-gen'ing and throw it out.

This made more sense when the data was consumed by NGEN, which needed to happen on the end-user's machine, but since cross-gen happens during the build it feels like an unnecessary extra step.  Cross-gen should be able to read the IBC data separately.

One could still say IBCMerge has some value for NuGet packages where we aren't cross-gen'ing before we ship but even in those cases, our use of IBC data is out of sync with how we build the product:
https://github.com/dotnet/runtime/blob/05954773b6ac44d171dded50e79686619fa71522/eng/codeOptimization.targets#L19-L20

We only gather IBC data for two different runs environments: Windows and Linux, but then arbitrarily choose which one to use based on the overall TargetOS of the build, not the target of an individual library.  So in some cases we don't ship OS specific configurations of the library but we embed crossgen data from some OS specific run.  In other cases we build OS specific configurations of the library that are more granular than just Windows/Linux.  Not to even get into the fact that many of our libraries cross-target by framework and that isn't even encoded in the IBC data we use today which is only collected for latest .NETCore.

Would it make sense to rethink how IBC Data flows through the system?  At the very least it seems like we can eliminate the embedding step for dotnet/runtime.  Perhaps we could improve the mapping between library and data as well with something like a deterministic identifier that could be used to represent the target of the library (akin to the TargetFramework strings we use in the build).

/cc @brianrob @davidwrighton

**Comment by davidwrighton** [comment]
Yes, everything in here makes sense. I'm actually in the process of open sourcing dotnet-pgo which is envisioned as a replacement for ibcmerge and .ibc data. Unfortunately, the data it produces is only consumable from crossgen2, so actual replacement of ibcmerge will need to wait for a post .NET 5.0 timeframe, but the general model is that pgo data will be passed into the crossgen2 compiler directly, instead of being embedded at IL build time. (Also the data is held in containers which are associated with a scenario instead of with a particular binary, thus changing how data proliferates through the system.)

**Comment by mangod9** [comment]
@davidwrighton is this issue still relevant with all the new pgo work in 6?

**Comment by davidwrighton** [comment]
We no longer use IBCMerge, and have switched to dotnet-pgo.
